### PR TITLE
Reserve an error message for "obj was inferred"

### PR DESCRIPTION
**Pull request by Smaug123**
This is boilerplate around https://github.com/fsharp/fslang-suggestions/issues/696 .

The actual implementation is https://github.com/dotnet/fsharp/pull/13298, but that has rotted several times. I'd like to get this boilerplate out of the way so that it stops rotting.

### CODE REVIEW

File: src/Compiler/FSComp.txt [hunk 1]
```diff
@@ -1672,4 +1672,5 @@ featureEscapeBracesInFormattableString,"Escapes curly braces before calling Form
...
 3556,chkImplementingInterfacesOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Implementing interfaces is not allowed."
 3557,chkAbstractMembersDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Abstract member declarations are not allowed."
-3558,chkExplicitFieldsDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Explicit field declarations are not allowed."
\ No newline at end of file
+3558,chkExplicitFieldsDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Explicit field declarations are not allowed."
+3559,typrelNeverRefinedAwayFromTop,"A type inference variable has been implicitly inferred to have type `obj`. Consider adding explicit type annotations. This warning is off by default and has been explicitly enabled for this project. You may suppress this warning by using #nowarn \"3559\"."
```
Comment (psfinaki) [inline src/Compiler/FSComp.txt:1676]: I believe the last but one sentence can become untrue with time and the last sentence is generally applicable to whatever warning - so I'd remove those for brevity.

Comment (Smaug123) [reply src/Compiler/FSComp.txt]: This wording was Don's (https://github.com/dotnet/fsharp/pull/13298#discussion_r899036766) - I do agree with you. Is the new version better?

Comment (psfinaki) [reply src/Compiler/FSComp.txt]: Hmhm not sure. Honestly I don't think it's good to suggest disabling the warning in any scenario but I might not have enough context here. @abelbraaksma?

Comment (abelbraaksma) [reply src/Compiler/FSComp.txt]: Disabling a warning as a suggestion in a warning is common practice, quite a few warnings already have that hint. I wouldn’t be too worried about making that part of the message.

Though, about wording, “a type inference variable” is not really a thing. Though I understand its meaning, there may not be a variable at all (like, try `Ok “”`, the Error type ends up `obj`, but won’t have a variable).

Perhaps: “a type inference variable or free type parameter”? (Or without the word “free”)

Comment (Smaug123) [reply src/Compiler/FSComp.txt]: Honestly just "a type" seems unambiguous to me.

File: tests/FSharp.Test.Utilities/Compiler.fs [hunk 2]
```diff
@@ -397,14 +397,14 @@ module rec Compiler =
 
     let withLangVersionPreview (cUnit: CompilationUnit) : CompilationUnit =
         withOptionsHelper [ "--langversion:preview" ] "withLangVersionPreview is only supported on F#" cUnit
-        
+
```
Comment (Smaug123) [inline tests/FSharp.Test.Utilities/Compiler.fs:400]: Changes to this file appear to have snuck in from my other branch, sorry, but I claim they are good changes.

**Comment by Smaug123** [comment]
Now "A type has been implicitly inferred as `obj`, which may be unintended" plus the "consider disabling".

**Comment by Smaug123** [comment]
I would like to get this in at some point in the near future, because this is the part of the change that rots the fastest.

**Review by psfinaki — APPROVED** [review]

**Comment by psfinaki** [comment]
I am fine with the wording now. @dotnet/fsharp-team-msft ping

File: src/Compiler/FSComp.txt (hunk 1 above)
Comment (abelbraaksma) [reply src/Compiler/FSComp.txt]: Good point, the word “inferred” is already in the sentence. 

> A type has been implicitly inferred ….

File: src/Compiler/FSComp.txt [hunk 3]
```diff
@@ -1672,4 +1672,5 @@ featureEscapeBracesInFormattableString,"Escapes curly braces before calling Form
 3555,chkInstanceLetBindingOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Instance let bindings are not allowed."
 3556,chkImplementingInterfacesOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Implementing interfaces is not allowed."
 3557,chkAbstractMembersDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Abstract member declarations are not allowed."
-3558,chkExplicitFieldsDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Explicit field declarations are not allowed."
\ No newline at end of file
+3558,chkExplicitFieldsDeclarationsOnStaticClasses,"If a type uses both [<Sealed>] and [<AbstractClass>] attributes, it means it is static. Explicit field declarations are not allowed."
+3559,typrelNeverRefinedAwayFromTop,"A type has been implicitly inferred as `obj`, which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with `#nowarn \"3559\"`."
```
Comment (abelbraaksma) [inline src/Compiler/FSComp.txt:1676]: Don't use backticks in the messages. See, for instance, line 1658 above.

```suggestion
3559,typrelNeverRefinedAwayFromTop,"A type has been implicitly inferred as 'obj', which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with '#nowarn \"3559\"'."
```

File: src/Compiler/xlf/FSComp.txt.cs.xlf [hunk 4]
```diff
@@ -1137,6 +1137,11 @@
         <target state="translated">Rozhraní {0} nemůžete implementovat se dvěma instancemi {1} a {2}, protože by se mohly sjednotit.</target>
         <note />
       </trans-unit>
+      <trans-unit id="typrelNeverRefinedAwayFromTop">
+        <source>A type has been implicitly inferred as `obj`, which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with `#nowarn \"3559\"`.</source>
+        <target state="new">A type has been implicitly inferred as `obj`, which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with `#nowarn \"3559\"`.</target>
+        <note />
+      </trans-unit>
```
Comment (abelbraaksma) [inline src/Compiler/xlf/FSComp.txt.cs.xlf:1144]: Don't use backticks in the messages. See, for instance, line 1658 above.

```suggestion
      <trans-unit id="typrelNeverRefinedAwayFromTop">
        <source>A type has been implicitly inferred as 'obj', which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with '#nowarn \"3559\"'.</source>
        <target state="new">A type has been implicitly inferred as `obj`, which may be unintended. Consider adding explicit type annotations, disabling the warning, or suppressing it locally with '#nowarn \"3559\"'.</target>
        <note />
      </trans-unit>
```

**Review by abelbraaksma — CHANGES_REQUESTED** [review]
Small nitpick, see below (I didn't repeat the comment for all translation files, these are generated automatically anyway). Otherwise good.

File: src/Compiler/FSComp.txt (hunk 3 above)
Comment (abelbraaksma) [reply src/Compiler/FSComp.txt:1676]: Actually, I think it is better to be consistent with the other wording used above. Consistency across the board is better. 

I mean:

```suggestion
3559,typrelNeverRefinedAwayFromTop,"A type has been implicitly inferred as `obj`, which may be unintended. Consider adding explicit type annotations. You can disable this warning by using '#nowarn \"3535\"' or '--nowarn:3535'."
```

**Review by abelbraaksma — APPROVED** [review]
Looking good! I think this can be merged @vzarytovskii, @dsyme, @KevinRansom

**Review by T-Gro — APPROVED** [review]

**Pull request merged**